### PR TITLE
AO3-6143 Prevent protected users from receiving gifts

### PR DESCRIPTION
**Pull request by sarken**
## Issue

https://otwarchive.atlassian.net/browse/AO3-6143

## Purpose

A NICE TRIP TO THE FOREST.

By which I mean, some fool said, "Until we get a preference in place for users to turn gifts on and off, we can easily use the new protected user role to help anyone who writes in with a complaint they're getting unwanted gifts." That fool was me. I am the fool who assumed the check was going to be the easy part.

Please see 75 million comments in the code about why this is the way it is instead of the nice simple approach I thought I could do in the initial commit.

I've got the obvious tests passing, so let's open this pull request and see what other test failures there are 🙃 

## Testing Instructions

Refer to Jira.

### CODE REVIEW

**Comment by tickinginstant** [comment]
I'm not sure it's a good idea to run the check on every gift every time the work is saved. That means that every time someone is added to the protected users list, any work(s) previously gifted to them will immediately become invalid, which may e.g. cause issues with collection reveals.

I think it also causes any gifts rejected by a user who is later added to the protected users list to become uneditable (because the only way the creator can edit the work is to remove the invalid recipient, but they can't remove the recipient once the gift has been rejected).

(Also, it seems like it'd be cleaner to refactor the gifts code to use `gifts.build` over `new_recipients`, instead of writing the same check twice. Particularly since it sounds like there'll be more gift validity checks added in the near future, which will presumably run into the same issues.)

**Comment by zz9pzza** [comment]
Would it make sense to add a property saying this person is not allowed to give things. 

They could make more sock puppets but it feels more natural. The ability to give a gift is  a right that could be removed ?

**Comment by sarken** [comment]
No, PAC specifically doesn't want the ability to turn off individuals' powers to do specific things -- plus it leaves them playing whack-a-mole with sock puppets, as you mentioned.

**Comment by sarken** [comment]
@tickinginstant Ugh, good point about the rejected gifts and challenge reveals. We'll just have to live with gifts staying on and folks will have to reject any troll/spam gifts.

>Also, it seems like it'd be cleaner to refactor the gifts code to use gifts.build over new_recipients, instead of writing the same check twice.

Just to clarify, do you mean build the gifts for the work in `recipients=` so the gifts rather than the list of recipients gets handed off to the validation? Then weed out gifts that are invalid (for any reason) in the validation?

**Comment by sarken** [comment]
@tickinginstant Oh, wait, you mean... everything in here related to gifts, so we can eliminate `new_recipients` in favor of something like `new_gifts`. 😬 That's a bit not in the spirit of a quick emergency fix, but I'll give it a shot.

**Comment by sarken** [comment]
"Sarken, did you even run the tests locally?" you might find yourself asking. 

I, however, am unable to reply as I have hurled my laptop against the wall and shattered it into a million pieces after seeing every test I ran (except the work model spec, I will accept that victory) turn red once the code has been pushed.

**Comment by sarken** [comment]
Closing, since this hasn't been merged and we're ready to do the preference. https://otwarchive.atlassian.net/browse/AO3-6257

File: app/models/work.rb [hunk 1]
```diff
@@ -444,40 +468,33 @@ def recipients=(recipient_names)
...
-          g.save
-        end
-      end
+  def save_new_gifts
+    return if self.new_gifts.blank?
+
+    self.new_gifts.each do |gift|
+      next if self.gifts.for_name_or_byline(gift.recipient).present?
+
+      Gift.create(recipient: gift.recipient, work: self)
```
[the remark this answers is not in the document]
Comment (redsummernight) [reply app/models/work.rb:489]: I see the same line in #4110. Do you remember why it is necessary to recreate the gift here?

[the remark this answers is not in the document]
Comment (redsummernight) [reply app/models/work.rb:489]: Answered in https://github.com/otwcode/otwarchive/pull/4110#discussion_r766211129, thanks!

**Pull request closed without merging**